### PR TITLE
[WIP] Change to GenericHost

### LINKED ISSUE #111: Implement GenericHost
In an effort to streamline and standardize DI, logging, and the lifetime of the bot, I would like to propose changing to a GenericHost.

Items to discuss:

* [ ] Replace Dependency Injection with `Microsoft.Extensions.DependencyInjection` 
* [ ] Design server runtime around a GenericHost
  * [ ] Use Console Lifetime (naturally handles CTRL+C or SIGTERM and calls `IHostApplicationLifetime.StopApplication()` to gracefully stop the bot as well as unblocking asynchronous runtimes). Works very well with `systemd`'s service management on the Linux side as well.
  * [ ] Lifetime managed by `IApplicationLifetime` and `CancellationToken`s
* [ ] Use `Microsoft.Extensions.Logging` base (compatible with existing logging library).
* [ ] App Configuration (environment variables, appsettings.json, UserSecrets)

For those unfamiliar, the [GenericHost](https://docs.microsoft.com/en-us/dotnet/core/extensions/generic-host) is based off the `WebHost` class from ASP.NET Core. However, this has the web-specific features removed in favor of a more generic long-lived service lifetime. The `HostBuilder` provides a set of utilities for configuring the host, such as reading configuration items, setting up database connections, and registering services. The Obsidian server itself is registered as a [BackgroundService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.backgroundservice?view=dotnet-plat-ext-5.0) which is responsible for handling startup (database migrations and so forth).

```cs
public override async Task ExecuteAsync(CancellationToken stoppingToken)
{
    // Set up server

    Console.WriteLine("Obsidian started successfully!");

    stoppingToken.WaitHandle.WaitOne(); // Wait for a stop to be requested by the `IApplicationLifetime`.

    Console.WriteLine("Obsidian shutting down...");

    // Tear down

    Console.WriteLine("Obsidian stopped.");
}
```

If we decide to move forward with this, I would be more than happy to be responsible for developing this and submitting a PR (or several).

Edit: I should point out that this will likely be a disruptive change since it involves redesigning the core of the bot as well as its lifetime management. However, I do believe that this will result in aiding future development and promote contributions by providing a well-known DI system.

### PR DESCRIPTION
**Pull request by TheBrambleShark**
Fixes #111 

* [x] GenericHost as bot lifetime
* [x] Reimplement multiple servers
* [ ] Deprecate `Obsidian.Globals` class in favor of injected services.
  * [ ] `Globals.HttpClient` -> `HttpService`
  * [ ] `Globals.Random` -> `RandomService`
  * [ ] `Globals.Config` -> `IOptions<GlobalConfig>`
  * [ ] `Globals.PacketLogger` -> `ILogger<T>`
  * [ ] `Globals.JsonSettings` -> `IOptions<JsonSerializerSettings>`
* [ ] Convert static helper classes into injected services, either singletons or scoped (singleton for that scope, at least until that scope is disposed)
  * [ ] `Obsidian.Utilities.Registry` -> `Obsidian.Utilities.RegistryService`
* [ ] Convert some models into database models
  * [ ] Server
* [ ] General testing

### CODE REVIEW

File: Obsidian/Obsidian.csproj [hunk 1]
```diff
@@ -52,15 +53,22 @@
...
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.8.0" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="5.0.1" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="5.0.0" />
+    <PackageReference Include="MySql.EntityFrameworkCore" Version="5.0.3.1" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="5.0.6" />
     <PackageReference Include="sharpcompress" Version="0.27.1" />
     <PackageReference Include="SharpNoise" Version="0.12.1.1" />
     <PackageReference Include="SharpZipLib" Version="1.3.1" />
+    <PackageReference Include="System.Text.Json" Version="5.0.2" />
```
Comment (TheBrambleShark) [inline Obsidian/Obsidian.csproj:72]: Seems this was left in-

I added the library to do a quick comparison but it seems it kept the package ref even though I uninstalled it. I'll remove it manually. Moving to S.T.J may be worth considering but is outside of the scope of this change.

**Review by Craftplacer — DISMISSED** [review]
Seems good to far.

I haven't looked at the entire source tree since *quite a while* (which I should really do at some point) but is there a reason why you rename `ObsidianServer` to `ObsidianCore`? (looking around in that code it already seems confusing enough that we have a property of type `Server` inside it)

**Comment by TheBrambleShark** [comment]
> 
> 
> Seems good to far.
> 
> I haven't looked at the entire source tree since _quite a time_ (which I should really do at some point) but is there a reason why you rename `ObsidianServer` to `ObsidianCore`? (looking around in that code it already seems confusing enough that we have a property of type `Server` inside it)

As for the rename, I felt having both `Server` and `ObsidianServer` types added unnecessary ambiguity. Renaming `ObsidianServer` to `ObsidianCore` better describes the function of its class, that being the core runtime of the application.

The `Server` property was removed in a subsequent push. It had originally been there as a way to have a single-server environment, but I had subsequently re-implemented the multi-server functionality.

**Comment by Naamloos** [comment]
stale PR, closing since changes would need to be redone due to changes in the main project.

**Pull request closed without merging**